### ISSUE TITLE
[Feature] tt-moe proposal (updated scope for Moe only)

### ISSUE THREAD

**Issue by ntarafdar** [tenstorrent/tt-metal#37456]
## **Problem Statement**

Currently, DeepSeek-V3 and GPT-OSS maintain separate MoE implementations with significant architectural differences:

**DeepSeek-V3:**
- MoEGate router with score correction bias and expert scaling
- SharedExpert runs in parallel with MoE experts
- All-gather/reduce-scatter for tensor parallelism at the decoder block level
- All-to-all dispatch/combine operations embedded within MoE module

**GPT-OSS:**
- TopKRouter with simple linear + topk + softmax logic
- Two expert modes: standard (sparse matmul) and throughput (all-to-all)
- No shared expert component
- Different collective operation patterns

This separation causes:
- Code duplication for similar MoE functionality
- Inability to test different configurations easily
- Maintenance burden for two separate implementations
- Difficulty comparing architectures side-by-side
- No clear path to support other MoE models (Mixtral, Kimi K2, Switch Transformer, etc.)

## **Proposed Solution**

Create a single, configurable `MoEBlock` module that can be configured via JSON to support both DeepSeek-V3 and GPT-OSS architectures, with extensibility for other MoE models. This is a focused scope reduction from the full tt-moe library proposed in #37354.

## **Architecture**

```
models/demos/unified_moe/
├── moe_block.py           # Main unified MoE block
├── configs/
│   ├── deepseek_v3.json   # DeepSeek-V3 configuration
│   ├── gpt_oss.json       # GPT-OSS configuration
│   ├── mixtral.json       # Mixtral 8x7B configuration
│   ├── switch_transformer.json  # Switch Transformer config
│   ├── arctic.json        # Arctic residual MoE config
│   └── kimi_k2.json       # Hypothetical Kimi K2 config
├── components/
│   ├── routers/
│   │   ├── base_router.py
│   │   ├── topk_router.py      # GPT-OSS style
│   │   ├── moe_gate.py         # DeepSeek style with corrections
│   │   ├── switch_router.py    # Top-1 with capacity
│   │   └── soft_router.py      # Soft routing (all experts)
│   └── experts/
│       ├── base_expert.py
│       ├── distributed_expert.py  # With all-to-all
│       ├── standard_expert.py     # Sparse matmul
│       ├── shared_expert.py       # DeepSeek shared expert
│       ├── residual_expert.py     # Arctic-style with dense backbone
│       └── geglu_expert.py        # GeGLU activation variant
└── tests/
    ├── test_moe_block.py
    ├── test_routers.py
    ├── test_experts.py
    ├── test_configs.py
    └── test_multi_model.py
```

## **Key Design Principles**

1. **Simplified Tensor Parallelism**: TP always follows the pattern: all-gather → compute → reduce-scatter. Only the cluster axis needs configuration.

2. **Direct TTNN Calls**: All collective operations call ttnn directly, no abstraction layers.

3. **JSON-Driven Configuration**: All architectural choices controlled via JSON files.

4. **Extensible Design**: Easy to add new router types, expert architectures, and models.

### **Base Schema**
```json
{
  "moe_block": {
    "name": "model_name",
    "architecture_type": "sparse_moe",  // or "soft_moe", "residual_moe"
    
    "tensor_parallel": {
      "enabled": true,
      "cluster_axis": 1  // Which mesh dimension for TP (all-gather/reduce-scatter)
    },
    
    "router": {
      "type": "topk",  // or "moe_gate", "switch", "soft_router"
      "config": {
        "num_experts": 8,
        "num_experts_per_tok": 2,
        "hidden_size": 4096,
        // Router-specific parameters
      }
    },
    
    "experts": {
      "distributed": {
        "enabled": true,
        "dispatch_cluster_axis": 0,  // Which mesh dimension for EP (all-to-all)
        "intermediate_size": 14336,
        "hidden_size": 4096,
        "num_experts": 8,
        "activation": "swiglu"  // or "gelu", "geglu", "silu"
      },
      "shared": {
        "enabled": false,
        "intermediate_size": 1408,
        "parallel_with_moe": true
      },
      "dense_backbone": {  // For residual MoE architectures
        "enabled": false,
        "intermediate_size": 8192,
        "residual_weight": 0.5
      }
    }
  }
}
```

### **DeepSeek-V3 Configuration**
```json
{
  "moe_block": {
    "name": "deepseek_v3_moe",
    "architecture_type": "sparse_moe",
    
    "tensor_parallel": {
      "enabled": true,
      "cluster_axis": 1  // TP on columns of mesh
    },
    
    "router": {
      "type": "moe_gate",
      "config": {
        "num_experts": 160,
        "num_experts_per_tok": 6,
        "n_routed_experts": 160,
        "n_group": 8,
        "topk_group": 3,
        "score_correction_bias": true,
        "routed_scaling_factor": 1.0,
        "memory_config": "L1_MEMORY_CONFIG"
      }
    },
    
    "experts": {
      "distributed": {
        "enabled": true,
        "dispatch_cluster_axis": 0,  // EP on rows of mesh
        "memory_config": "L1_MEMORY_CONFIG",
        "intermediate_size": 10944,
        "hidden_size": 5120,
        "num_experts_per_device": 20,
        "activation": "swiglu"
      },
      "shared": {
        "enabled": true,
        "intermediate_size": 1408,
        "hidden_size": 5120,
        "parallel_with_moe": true,
        "memory_config": "L1_MEMORY_CONFIG"
      },
      "dense_backbone": {
        "enabled": false
      }
    }
  }
}
```

### **GPT-OSS Configuration**
```json
{
  "moe_block": {
    "name": "gpt_oss_moe",
    "architecture_type": "sparse_moe",
    
    "tensor_parallel": {
      "enabled": false  // GPT-OSS doesn't use TP at MoE level
    },
    
    "router": {
      "type": "topk",
      "config": {
        "num_experts": 128,
        "num_experts_per_tok": 8,
        "hidden_size": 6144,
        "memory_config": "DRAM_MEMORY_CONFIG"
      }
    },
    
    "experts": {
      "distributed": {
        "enabled": true,
        "dispatch_cluster_axis": 0,  // EP dispatch axis
        "memory_config": "L1_MEMORY_CONFIG",
        "intermediate_size": 20480,
        "hidden_size": 6144,
        "num_experts": 128,
        "weight_dtype": "bfloat4_b",
        "activation": "swiglu",
        "swiglu_limit": 32768
      },
      "shared": {
        "enabled": false
      },
      "dense_backbone": {
        "enabled": false
      }
    }
  }
}
```

### **Mixtral Configuration**
```json
{
  "moe_block": {
    "name": "mixtral_8x7b",
    "architecture_type": "sparse_moe",
    
    "tensor_parallel": {
      "enabled": false
    },
    
    "router": {
      "type": "topk",
      "config": {
        "num_experts": 8,
        "num_experts_per_tok": 2,
        "hidden_size": 4096,
        "memory_config": "DRAM_MEMORY_CONFIG"
      }
    },
    
    "experts": {
      "distributed": {
        "enabled": true,
        "dispatch_cluster_axis": 0,
        "intermediate_size": 14336,
        "hidden_size": 4096,
        "num_experts": 8,
        "activation": "swiglu"
      },
      "shared": {
        "enabled": false
      },
      "dense_backbone": {
        "enabled": false
      }
    }
  }
}
```

### **Kimi K2 Configuration (Hypothetical)**
```json
{
  "moe_block": {
    "name": "kimi_k2",
    "architecture_type": "sparse_moe",
    
    "tensor_parallel": {
      "enabled": true,
      "cluster_axis": 1
    },
    
    "router": {
      "type": "moe_gate",
      "config": {
        "num_experts": 256,
        "num_experts_per_tok": 6,
        "score_correction_bias": true,
        "routed_scaling_factor": 1.0,
        "load_balancing": {
          "enabled": true,
          "auxiliary_loss_weight": 0.01,
          "z_loss_weight": 0.001
        }
      }
    },
    
    "experts": {
      "distributed": {
        "enabled": true,
        "dispatch_cluster_axis": 0,
        "num_experts_per_device": 32,
        "activation": "swiglu"
      },
      "shared": {
        "enabled": true,
        "parallel_with_moe": true
      }
    }
  }
}
```

### **Switch Transformer Configuration**
```json
{
  "moe_block": {
    "name": "switch_transformer",
    "architecture_type": "sparse_moe",
    
    "tensor_parallel": {
      "enabled": false
    },
    
    "router": {
      "type": "switch",
      "config": {
        "num_experts": 2048,
        "routing_strategy": {
          "type": "top_1",
          "capacity_factor": 1.25,
          "drop_tokens": true
        }
      }
    },
    
    "experts": {
      "distributed": {
        "enabled": true,
        "dispatch_cluster_axis": 0,
        "capacity_constrained": true
      }
    }
  }
}
```

### **Arctic Configuration (Residual MoE)**
```json
{
  "moe_block": {
    "name": "arctic",
    "architecture_type": "residual_moe",
    
    "tensor_parallel": {
      "enabled": false
    },
    
    "router": {
      "type": "topk",
      "config": {
        "num_experts": 128,
        "num_experts_per_tok": 2
      }
    },
    
    "experts": {
      "distributed": {
        "enabled": true,
        "dispatch_cluster_axis": 0,
        "intermediate_size": 4864,
        "residual_connection": true
      },
      "dense_backbone": {
        "enabled": true,
        "intermediate_size": 8192,
        "residual_weight": 0.5,
        "parallel_with_moe": true
      }
    }
  }
}
```

**Comment by ntarafdar** [comment]
## **Implementation Details**

### **Unified MoEBlock Class**
```python
class MoEBlock:
    def __init__(self, config_path: str, mesh_device, ccl):
        self.config = load_json_config(config_path)
        self.mesh_device = mesh_device
        self.ccl = ccl
        
        # Extract TP config
        self.tp_config = self.config.get("tensor_parallel", {"enabled": False})
        self.tp_enabled = self.tp_config.get("enabled", False)
        
        if self.tp_enabled:
            self.tp_axis = self.tp_config["cluster_axis"]
            self.tp_size = mesh_device.shape[self.tp_axis]
            
        # Initialize components based on config
        self._init_router()
        self._init_experts()
    
    def forward(self, x, mode="decode"):
        """
        Forward pass with simplified collective operations.
        TP pattern: all-gather → compute → reduce-scatter
        EP pattern: all-to-all dispatch → compute → all-to-all combine
        """
        # Step 1: All-gather if TP is enabled and input is sharded
        if self.tp_enabled and self._is_tp_sharded(x):
            x = ttnn.experimental.all_gather_async(
                x,
                cluster_axis=self.tp_axis,
                dim=-1,
                memory_config=self._get_memory_config(mode, "all_gather"),
                topology=ttnn.Topology.Linear
            )
            x_was_gathered = True
        else:
            x_was_gathered = False
        
        # Step 2: Router
        indices, weights = self.router(x)
        
        # Step 3: Expert computation (potentially parallel)
        outputs = []
        
        # Distributed experts with EP
        if self.distributed_expert:
            expert_out = self._run_distributed_experts(x, indices, weights, mode)
            outputs.append(expert_out)
        
        # Shared expert (if enabled and parallel)
        if self.shared_expert and self.config["experts"]["shared"]["parallel_with_moe"]:
            shared_out = self.shared_expert(x)
            outputs.append(shared_out)
        
        # Dense backbone (for residual MoE)
        if self.dense_backbone:
            dense_out = self.dense_backbone(x)
            outputs.append(dense_out)
        
        # Step 4: Combine outputs
        if len(outputs) > 1:
            output = sum(outputs)  # Or weighted sum for residual MoE
        else:
            output = outputs[0]
        
        # Step 5: Reduce-scatter if we did all-gather
        if x_was_gathered:
            output = ttnn.experimental.reduce_scatter_minimal_async(
                output,
                cluster_axis=self.tp_axis,
                dim=3,
                memory_config=self._get_memory_config(mode, "reduce_scatter"),
                topology=ttnn.Topology.Linear
            )
        
        return output
    
    def _run_distributed_experts(self, x, indices, weights, mode):
        """Run distributed experts with EP (different axis from TP)"""
        ep_axis = self.config["experts"]["distributed"]["dispatch_cluster_axis"]
        
        # All-to-all dispatch on EP axis
        dispatched = ttnn.experimental.all_to_all_dispatch(
            x, indices,
            cluster_axis=ep_axis,
            memory_config=self._get_memory_config(mode, "dispatch")
        )
        
        # Expert computation
        expert_out = self.distributed_expert(dispatched)
        
        # All-to-all combine on EP axis
        combined = ttnn.experimental.all_to_all_combine(
            expert_out, indices,
            cluster_axis=ep_axis,
            memory_config=self._get_memory_config(mode, "combine")
        )
        
        # Apply routing weights
        weighted = ttnn.multiply(combined, weights)
        
        return weighted
    
    def _is_tp_sharded(self, x):
        """Check if input is TP-sharded"""
        if not self.tp_enabled:
            return False
        
        hidden_size = self.config["experts"]["distributed"]["hidden_size"]
        actual_dim = x.shape[-1]
        expected_sharded_dim = hidden_size // self.tp_size
        
        return actual_dim == expected_sharded_dim
```

**Comment by ntarafdar** [comment]
## **Success Criteria**

- ✅ Single MoEBlock module supports both DeepSeek and GPT-OSS configurations
- ✅ JSON configuration files control all architectural choices
- ✅ PCC > 0.99 compared to existing implementations  
- ✅ No performance regression
- ✅ Clean separation of router and expert implementations
- ✅ Simplified collective operations (TP: just cluster axis, EP: dispatch axis)
- ✅ Support for additional MoE architectures (Mixtral, Switch Transformer, Arctic)
- ✅ Extensible design for future models (Kimi K2, etc.)

## **Benefits**

1. **Unified Testing**: Single module to test different MoE configurations
2. **Easy Experimentation**: Switch between architectures via JSON config
3. **Reduced Maintenance**: One codebase instead of multiple
4. **Clear Abstractions**: Separated router, expert, and collective op logic
5. **Future Extensibility**: Easy to add new router or expert types
6. **Model Comparison**: Side-by-side testing of different MoE approaches
7. **Simplified Configuration**: TP ops need only cluster axis specification

## **Out of Scope**

- Full tt-moe library implementation (deferred to future work)
- Attention mechanisms (MLA vs standard) 
- Complete decoder block unification
- Model-level integration
- Training-time features (gradients, optimizer states)

## **Notes**

- All TTNN collective operations are called directly without abstraction layers
- Tensor parallelism always follows: all-gather → compute → reduce-scatter pattern
- Expert parallelism uses all-to-all dispatch/combine on a different axis
- Memory configs and other parameters have sensible defaults based on mode (decode/prefill)

---

This issue creates a focused, configurable MoE block that unifies DeepSeek-V3 and GPT-OSS implementations while providing extensibility for other MoE architectures through JSON configuration.
